Remove keypress.js and move to an KeyboardEvent.code-based virtual cube controller for more robust input handling and keyboard-layout-agnostic UI

xeniagda opened the pull request:
Hi!

The `keypress.js`-library does not seem to be actively maintained, and it is based on the deprecated [`KeyboardEvent.keyCode`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) attribute which has a number of issues, the main one being that it does not work well across different keyboard layouts.

This pull request removes the `controls.js` dependency and introduces a simpler and more robust input handling interface, based on the recommended `event.code`, as well as checking `event.alt`/`event.ctrl`/`event.shift` for modifiers instead of tracking modifiers in a global state.

The pull request also updated `controls.html` to give the user the ability to change the key bindings to whatever they may like. The bindings are stored in `document.localStorage`. It only allows the user to bind keys to doing moves / algs, using space, enter, etc. to do other actions is still hard coded.

Commit:
Read keymaps from localStorage

Commit:
controls.html to edit controls

Commit:
Upgrade controls.html a bit

Commit:
Change control system

Commit:
Update controls.html to use new kb interface

Commit:
Remove controls.js from readme since we don't use that anymore

Commit:
Style index.html and controls.html somewhat

tao-yu commented:
Wow, this is some good work! I've always wanted to have this feature in the trainer 😮 . Thank you so much!

I've tested it out and it seems to work great. I would just ask for you to fix just one thing, after which I'm very happy to merge:

In the default keymap, it should be

Shift-H = S
Shift-G = S'

Where currently it is Shift-H = S F' and Shift-G = S' F

If you'd like, you can also change the link to the controls on index.html to say something like "View/Edit keyboard controls".

Again I can't tell you how grateful I am to you for doing this. In addition to it being a good feature, I have found keypress.js to cause some issues with certain adblockers in the past. so this fixes a lot of issues at once.

Commit:
Change link to controls

Commit:
Fix S-H and S-G

Commit:
Update 1LLL.html

xeniagda commented:
No problems at all! Thank you for making the trainer in the first place, it's been invaluable to me lately!

I fixed the default shortcuts with Shift-H and Shift-G and changed the link text. Also I updated the 1LLL.html to use the new control system, which I forgot before. It should be good to merge now.

tao-yu commented:
Awesome, everything looks good. I will merge it now.
